ancilla.el

```diff
@@ -106,57 +112,70 @@ call 'ancilla-rewrite' otherwise."
-(defun ancilla--call-adaptor-with-instruction (mode)
-  "Call the AI-powered adaptor with the given MODE and instruction.
-
-MODE should be either 'rewrite or 'generate.  The function prompts
-the user for an instruction, gets the current buffer context, and
-calls the appropriate function in the adaptor.  If the adaptor
-returns a suggestion, the function applies the suggestion to the
-buffer.
-
-If `ancilla-show-confirmation' is set to 't', the function shows a
-confirmation message before applying the suggestion.
-
-If `ancilla-show-confirmation' is set to 'rewrite-only', the
-function shows a confirmation message only when MODE is 'rewrite'.
-
-The function also calls any hooks registered with the
-`ancilla-hooks' property of the adaptor.  Each hook is called with
-a plist containing the keys :instruction, :buffer-context, and
-:mode."
-  (let* ((instruction
-          (if (eq mode 'ask)
-              (read-string "Question: ")
-            (read-string "Instruction: ")))
-         (buffer-context (ancilla--get-buffer-context))
-         (adaptor-request-fn-name
-          (cond
-           ((eq mode 'rewrite) 'ancilla-rewrite)
-           ((eq mode 'generate) 'ancilla-generate)
-           ((eq mode 'ask) 'ancilla-ask)))
+
+(defun ancilla--process-instruction (instruction mode)
+  "Process the given INSTRUCTION with the specified MODE."
+  (let* ((buffer-context (ancilla--get-buffer-context))
+         (adaptor-request-fn-name (ancilla--get-adaptor-function mode))
          (adaptor-request-fn (get ancilla-adaptor adaptor-request-fn-name))
-         (request-callback-fn
-          (cond
-           ((memq mode '(rewrite generate))
-            (apply-partially 'ancilla--diff-replace-selection
-                             mode
-                             (plist-get buffer-context :excursion)
-                             (plist-get buffer-context :selection)))
-           (t 'ignore)
-           ;; (t (lambda (&rest _)
-           ;;      (switch-to-buffer "*ancilla-chat*")))
-           )))
+         (request-callback-fn (ancilla--create-callback-function mode buffer-context)))
     (when-let ((hooks (get ancilla-adaptor 'ancilla-hooks)))
-      (dolist (hook hooks) (funcall hook
-                                    :instruction instruction
-                                    :buffer-context buffer-context
-                                    :mode mode)))
+      (ancilla--run-hooks hooks instruction buffer-context mode))
     (funcall adaptor-request-fn
              :instruction instruction
              :buffer-context buffer-context
              :callback request-callback-fn)))
 
+(defun ancilla--process-ask (instruction)
+  "Process the given INSTRUCTION for 'ask' mode."
+  (ancilla--process-instruction instruction 'ask))
+
+(defun ancilla--process-generate (instruction)
+  "Process the given INSTRUCTION for 'generate' mode."
+  (ancilla--process-instruction instruction 'generate))
+
+(defun ancilla--process-rewrite (instruction)
```

Comment:
this function can provide a shortcut for given fixed rewrite instruction (maybe it is possible)